File: backup/Twitch/app/src/main/java/com/exact/twitch/ui/player/clip/ClipPlayerViewModel.java
package com.exact.twitch.ui.player.clip;

import android.app.Application;
import android.net.Uri;
import android.text.format.DateUtils;

import com.exact.twitch.model.OfflineVideo;
import com.exact.twitch.model.clip.Clip;
import com.exact.twitch.model.clip.ClipStatusResponse;
import com.exact.twitch.repository.PlayerRepository;
import com.exact.twitch.ui.OnQualityChangeListener;
import com.exact.twitch.ui.player.PlayerViewModel;
import com.exact.twitch.util.DownloadUtils;
import com.exact.twitch.util.PlayerUtils;
import com.exact.twitch.util.TwitchApiHelper;
import com.google.android.exoplayer2.offline.ProgressiveDownloadAction;
import com.google.android.exoplayer2.source.ExtractorMediaSource;
import com.google.android.exoplayer2.source.MediaSource;
import com.google.android.exoplayer2.upstream.cache.CacheDataSourceFactory;

import java.util.Calendar;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.Map;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ClipPlayerViewModel extends PlayerViewModel implements OnQualityChangeListener {

    private Clip clip;
    private ExtractorMediaSource.Factory factory;
    private long playbackProgress;

    @Inject
    public ClipPlayerViewModel(Application context, PlayerRepository playerRepository) {
        super(context, playerRepository);
    }

    @Override
    public void changeQuality(int index) {
        playbackProgress = getPlayer().getCurrentPosition();
        play(getUrls().get(getQualities().getValue().get(index)));
    }

    @Override
    public boolean isFirstLaunch() {
        return clip == null;
    }

    public void setClip(Clip clip) {
        this.clip = clip;
    }

    public void play() {
        getQualities().setValue(new LinkedList<>());
        playerRepository.fetchClipQualities(clip.getSlug(), new Callback<ClipStatusResponse>() {
            @Override
            public void onResponse(Call<ClipStatusResponse> call, Response<ClipStatusResponse> response) {
                if (response.isSuccessful()) {
                    Map<CharSequence, String> urls = new HashMap<>();
                    factory = new ExtractorMediaSource.Factory(new CacheDataSourceFactory(DownloadUtils.INSTANCE.getCache(getApplication()), getDataSourceFactory()));
                    for (ClipStatusResponse.QualityOption qualityOption : response.body().getQualityOptions()) {
                        urls.put(qualityOption.getQuality(), qualityOption.getSource());
                        getQualities().getValue().add(qualityOption.getQuality());
                    }
                    setUrls(urls);
                    play(response.body().getQualityOptions().get(0).getSource());

                }
            }

            @Override
            public void onFailure(Call<ClipStatusResponse> call, Throwable t) {

            }
        });
//            if (clip.getVod() != null) {
//                initMessages();
//                fetchSubscriberBadges(Integer.parseInt(clip.getBroadcaster().getId()));
//            }
    }

    private void play(String source) {
        MediaSource mediaSource = factory.createMediaSource(Uri.parse(source));
        getPlayer().prepare(mediaSource);
        getPlayer().seekTo(playbackProgress);
        getPlayer().setPlayWhenReady(true);
    }

    public void download(String quality) {
        String url = getUrls().get(quality);
        ProgressiveDownloadAction progressiveDownloadAction = new ProgressiveDownloadAction(Uri.parse(url), false, null, null);
        OfflineVideo video = new OfflineVideo(url, clip.getTitle(), clip.getBroadcaster().getName(), clip.getGame(), clip.getDuration().longValue(), TwitchApiHelper.INSTANCE.parseIso8601Date(getApplication(), clip.getCreatedAt()), DateUtils.formatDateTime(getApplication(), Calendar.getInstance().getTime().getTime(), DateUtils.FORMAT_NO_YEAR), clip.getThumbnails().getMedium(), clip.getBroadcaster().getLogo());
        PlayerUtils.INSTANCE.startDownload(getApplication(), progressiveDownloadAction, video);
    }
}

File: app/src/main/java/com/exact/twitch/ui/player/clip/ClipPlayerViewModel.kt
package com.exact.twitch.ui.player.clip

import android.app.Application
import android.net.Uri
import com.exact.twitch.db.VideosDao
import com.exact.twitch.model.OfflineVideo

import com.exact.twitch.model.clip.Clip
import com.exact.twitch.repository.PlayerRepository
import com.exact.twitch.ui.OnQualityChangeListener
import com.exact.twitch.ui.player.PlayerHelper
import com.exact.twitch.ui.player.PlayerType
import com.exact.twitch.ui.player.PlayerViewModel
import com.exact.twitch.util.DownloadUtils
import com.exact.twitch.util.PlayerUtils
import com.exact.twitch.util.TwitchApiHelper
import com.google.android.exoplayer2.offline.ProgressiveDownloadAction
import com.google.android.exoplayer2.source.ExtractorMediaSource
import com.google.android.exoplayer2.upstream.cache.CacheDataSourceFactory
import io.reactivex.rxkotlin.addTo
import kotlinx.coroutines.experimental.launch

import javax.inject.Inject

class ClipPlayerViewModel @Inject constructor(
        context: Application,
        private val playerRepository: PlayerRepository,
        private val dao: VideosDao) : PlayerViewModel(context, PlayerType.VIDEO), OnQualityChangeListener {

    lateinit var clip: Clip
    private val factory: ExtractorMediaSource.Factory = ExtractorMediaSource.Factory(CacheDataSourceFactory(DownloadUtils.getCache(context), dataSourceFactory))
    private var playbackProgress: Long = 0
    val helper = PlayerHelper()

    override fun changeQuality(index: Int, tag: String) {
        if (helper.selectedQualityIndex != index) {
            playbackProgress = player.currentPosition
            play(helper.urls[helper.qualities.value!![index]]!!)
        }
    }

    fun play() {
        playerRepository.fetchClipQualities(clip.slug)
                .subscribe({
                    println(it)
                    val qualities = ArrayList<CharSequence>(it.size)
                    it.forEach { option ->
                        qualities.add(option.quality)
                        helper.urls[option.quality] = option.source
                    }
                    play(helper.urls[qualities[helper.selectedQualityIndex]]!!)
                    helper.qualities.postValue(qualities)
                    if (clip.vod != null) {
                        playerRepository.fetchSubscriberBadges(clip.broadcaster.id.toInt())
                                .subscribe({ response ->

                                }, { t ->

                                })
                                .addTo(compositeDisposable)

                    }
                }, {

                })
                .addTo(compositeDisposable)
    }

    private fun play(source: String) {
        play(factory.createMediaSource(Uri.parse(source)))
        player.seekTo(playbackProgress)
    }

    fun download(quality: String) {
        val context = getApplication<Application>()
        val url = helper.urls[quality]!!
        val downloadAction = ProgressiveDownloadAction(Uri.parse(url), false, null, null)
        val uploadDate = TwitchApiHelper.parseIso8601Date(context, clip.createdAt)
        val currentDate = TwitchApiHelper.getCurrentTimeFormatted(context)
        OfflineVideo(url, clip.title, clip.broadcaster.name, clip.game, clip.duration.toLong(), currentDate, uploadDate, clip.thumbnails.medium, clip.broadcaster.logo).let {
            launch { dao::insert }
            PlayerUtils.startDownload(context, downloadAction, it)
        }
    }
}
